### PR TITLE
Add Table::saveMany().

### PR DESCRIPTION
**Pull request by ADmad**
Allows saving multiple records in a transaction. This is a pretty common requirement and would be nice to have in core itself.

I'll add tests if people like the idea.

### CODE REVIEW

**Comment by thinkingmedia** [comment]
Would be better if validation was performed on all records first before a save is attempted.

**Comment by lorenzo** [comment]
The kind of comments, like the one from @thinkingmedia is what has stopped me from implementing this function. Everyone will have a different preference of how the save should be done.

**Comment by thinkingmedia** [comment]
@lorenzo the code is run in a transaction which would rollback if validation fails, but I don't know if all data sources support transactions. @ADmad know more about that sort of stuff than me. So if he says this is safe. I'll go with it :)

**Comment by lorenzo** [comment]
I understand how it works... what I meant is that there a many variations of this code and it will be difficult to please everyone

**Comment by thinkingmedia** [comment]
I don't think so. All that matters is the function signature and consistency with other methods. Saving an array of entities is going to have the same outcome no matter who's alternate implementation is used.

How about this. Next time you add a feature. I'll up vote it. :)

**Comment by ADmad** [comment]
@lorenzo We can document the fact that atomic save is all this method does. If anyone wants anything fancier they can roll their own :)

**Comment by ADmad** [comment]
> I don't know if all data sources support transactions.

@thinkingmedia They don't, those using them are out of luck :stuck_out_tongue:. BTW @lorenzo knows about the ORM than anyone else, so his opinion weighs way more than mine :)

**Comment by thinkingmedia** [comment]
@ADmad so I basically just insulted him. lol, sorry.

**Comment by markstory** [comment]
> I don't know if all data sources support transactions

All of the relational databases we support, do support transactions.

File: tests/TestCase/ORM/TableTest.php [hunk 1]
```diff
@@ -2778,6 +2778,43 @@ public function testUpdateNoPrimaryButOtherKeys()
...
+    {
+        $entities = [
+            new Entity(['name' => 'admad']),
+            new Entity(['name' => 'dakota'])
+        ];
+
+        $table = TableRegistry::get('authors');
+        $result = $table->saveMany($entities);
+        $this->assertSame($entities, $result);
+        $this->assertTrue(isset($result[0]->id));
```
Comment (markstory) [inline tests/TestCase/ORM/TableTest.php]: Should you check that all the entities have `isNew()` set to false?

File: tests/TestCase/ORM/TableTest.php [hunk 2]
```diff
@@ -2778,6 +2778,43 @@ public function testUpdateNoPrimaryButOtherKeys()
...
+
+        $entities = $table->find()->all();
+        $entities->first()->name = 'admad';
+
+        $result = $table->saveMany($entities);
+        $this->assertSame($entities, $result);
+
+        $first = $table->find()->first();
+        $this->assertSame('admad', $first->name);
+    }
```
Comment (markstory) [inline tests/TestCase/ORM/TableTest.php]: You should have a test for when saving fails. Perhaps a validation error on the second record, and ensure that the first entity's state is correct?

Comment (ADmad) [inline tests/TestCase/ORM/TableTest.php]: Tried that. The state of 1st entity does get changed. `isNew()` returns false for it even though transaction was rolled back and entity wasn't actually persisted. So what should be done?

**Comment by ADmad** [comment]
No idea why postgres doesn't like using a resultset instead of array.

File: tests/TestCase/ORM/TableTest.php (hunk 2 above)
Comment (ADmad) [inline tests/TestCase/ORM/TableTest.php]: Okay I see how _procesSave() restores state, will do same.

**Comment by ADmad** [comment]
@markstory Updated as per your comments.

**Comment by markstory** [comment]
Rolling back isNew() and primary key values looks good.

**Comment by ADmad** [comment]
> Rolling back isNew() and primary key values looks good. 

That's exactly what [_processSave()](https://github.com/cakephp/cakephp/blob/master/src/ORM/Table.php#L1483) does, so it should be :)

Need someone with postgres experience help figure out why tests are failing on it.

**Comment by ADmad** [comment]
Fixed! The lack of implicit ordering in postgres always gets me.

**Comment by chinpei215** [comment]
Looks good to me. Nice work!

File: src/ORM/Table.php [hunk 3]
```diff
@@ -1619,10 +1619,14 @@ protected function _update($entity, $data)
     }
 
     /**
-     * Saves multiple records for a table.
+     * Persists multiple entities of a table.
+     *
+     * The records will be saved in a transaction which will be rolled back if
+     * any one of the records fails to save due to failed validation or datbase
```
Comment (markstory) [inline src/ORM/Table.php]: typo in database.

File: src/ORM/Table.php [hunk 4]
```diff
@@ -1619,6 +1619,45 @@ protected function _update($entity, $data)
     }
 
     /**
+     * Persists multiple entities of a table.
+     *
+     * The records will be saved in a transaction which will be rolled back if
+     * any one of the records fails to save due to failed validation or datbase
```
Comment (markstory) [inline src/ORM/Table.php]: Typo in database.

**Comment by markstory** [comment]
Looks good to me. Should this go into `3.next`?

**Comment by ADmad** [comment]
> Should this go into 3.next?

We have been adding new features in bugfix releases. Do you have any particular concerns for not doing so in this case?

**Comment by markstory** [comment]
Thinking about it more. No. This is a net new method so its ability to break applications is very low.

**Pull request merged**